onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -0,0 +1,562 @@
+    return onnxruntime::make_unique<Provider_OrtMemoryInfo_Impl>(name_, type_, device_ ? static_cast<Provider_OrtDevice_Impl*>(device_)->v_ : OrtDevice(), id_, mem_type_);
+  }
+
+  std::unique_ptr<Provider_KernelDefBuilder> KernelDefBuilder_Create() override {
+    return onnxruntime::make_unique<Provider_KernelDefBuilder_Impl>();
+  }
+
+  std::shared_ptr<Provider_KernelRegistry> KernelRegistry_Create() override {
+    return std::make_shared<Provider_KernelRegistry_Impl>();
+  }
+
+  std::unique_ptr<Provider_IndexedSubGraph> IndexedSubGraph_Create() override {
+    return onnxruntime::make_unique<Provider_IndexedSubGraph_Impl>();
+  }
+
+  Provider_AllocatorPtr CreateAllocator(Provider_DeviceAllocatorRegistrationInfo& info, OrtDevice::DeviceId device_id = 0) override {
+    DeviceAllocatorRegistrationInfo info_real{
+        info.mem_type, [&info](int value) { return std::move(static_cast<Provider_IDeviceAllocator_Impl*>(&*info.factory(value))->p_); },
+        info.max_mem};
+
+    return std::make_shared<Provider_IAllocator_Impl>(onnxruntime::CreateAllocator(info_real, device_id));
+  }
+
+  std::unique_ptr<Provider_IDeviceAllocator>
+  CreateCPUAllocator(std::unique_ptr<Provider_OrtMemoryInfo> memory_info) override {
+    return onnxruntime::make_unique<Provider_IDeviceAllocator_Impl>(onnxruntime::make_unique<CPUAllocator>(std::move(static_cast<Provider_OrtMemoryInfo_Impl*>(memory_info.get())->info_)));
+  };
+
+  Provider_AllocatorPtr
+  CreateDummyArenaAllocator(std::unique_ptr<Provider_IDeviceAllocator> resource_allocator) override {
+    return std::make_shared<Provider_IAllocator_Impl>(onnxruntime::make_unique<DummyArena>(std::move(static_cast<Provider_IDeviceAllocator_Impl*>(resource_allocator.get())->p_)));
+  };
+
+  std::unique_ptr<Provider_IExecutionProvider_Router> Create_IExecutionProvider_Router(Provider_IExecutionProvider* outer, const std::string& type) override {
+    return onnxruntime::make_unique<Provider_IExecutionProvider_Router_Impl>(outer, type);
+  };
+
+  logging::Logger* LoggingManager_GetDefaultLogger() override {
+    return const_cast<logging::Logger*>(&logging::LoggingManager::DefaultLogger());
+  }
+
+  void* HeapAllocate(size_t size) override { return new uint8_t[size]; }
+  void HeapFree(void* p) override { delete reinterpret_cast<uint8_t*>(p); }
+
+  bool CPU_HasAVX2() override {
+    return CPUIDInfo::GetCPUIDInfo().HasAVX2();
+  }
+
+  bool CPU_HasAVX512f() override {
+    return CPUIDInfo::GetCPUIDInfo().HasAVX512f();
+  }
+
+  void LogRuntimeError(uint32_t session_id, const common::Status& status, const char* file, const char* function, uint32_t line) override {
+    return ::onnxruntime::LogRuntimeError(session_id, status, file, function, line);
+  }
+
+} provider_host_;
+
+struct ProviderLibrary {
+  ProviderLibrary(const char* filename) {
+    Env::Default().LoadDynamicLibrary(filename, &handle_);
+    if (!handle_)
+      return;
```

Comment:
Is there a way to know what the error was with the loading of the DLL?

Reply:
Looks like we can call GetLastError() if it fails. What did you want to do on failure?

Reply:
Sure, that should be enough. Let's log the error.

Reply:
The LoadDynamicLibrary function returns a Status object. Can we use that to log the error?

Reply:
Added a log statement